### DISCUSSION TITLE
Feature request: optional two-dot diff view on pull requests

### DISCUSSION THREAD

**Discussion by jessiejofielding** [community/community#14528] [Pull Requests]
## Bottom Line Up Front
The default git diff is two dot, many other platforms use two dot, and whether or not it's necessary, it would certainly be nice to have. I want it, and I'm pretty sure lots of people want it.

If you love three dot diffs, awesome! Don't worry, my proposal doesn't want to take your diff preferences away.

## Proposal
Create an optional way to use two dot diffs in pull requests. Potentially have some note, tag, or other UI difference like color to show which kind of diff is being used.

#### Suggestion 1: 
make two dot diffs vs three dot diffs a repo setting

#### Suggestion 2: 
have a toggle upon creation of each pull request that allows you to choose two dot diffs vs three dot diffs

## Two dot vs three dot diffs, a brief background:
```    
      B (feature-branch) 
    /
- A ---C (target-branch)
```

#### Two dot
`git diff feature-branch target-branch` is the same as `git diff feature-branch..target-branch`. 
It will show the difference between B and C (i.e. the latest commits on both branches).

#### Three dot
`git diff feature-branch...target-branch` will show the difference between B and A (i.e. the latest commit on feature-branch and the most recent common ancestor of feature-branch and target-branch).
Pull request diffs on GitHub use this currently.

If you open a pull request `feature-branch -> target-branch`, you may want to know only what changes feature-branch implements (B compared to A), or you may want to know all the ways feature-branch differs from the branch it's about to merge into (B compared to C).

## Action items for fellow users

- Support this feature request if you'd like it! 
- Share it with your colleagues or friends or twitter feeds if you're so inclined :) 
- If you have a workaround you use regularly for this, wouldn't you rather work with GitHub instead of around it? If so, _**please consider supporting this feature request instead of explaining workarounds.**_ 
   - Listing workarounds can make the feature request less clear, and I'd imagine can make it hard for GitHub staff to tell whether the workaround is being suggested as a feasible solution in place of the feature request, or as a temporary patch.
- Share your technical expertise and implementation ideas in this discussion!

(Nothing will come of this if GitHub staff doesn't know how badly we want this, so let's show them)

For reference, the main workarounds I've seen are:
- to compare branches outside of the PR and edit the URL to have two dots instead of three, or 
- to merge target-branch into feature-branch/to rebase (this option works because it changes the most recent common ancestor to C, but is still a three dot diff)

## Thanks

@github-staff thanks for considering this! I hope it gains enough support to make it onto your radar. I wasn't sure how much detail to go into with the proposal -- don't want to take all the design work away from your team, but I'm happy to flesh the design out more if you'd like.

Everyone, thanks for reading this far! If you've shown support for this in some way or another, double thanks (I really like two dot diffs)

**Comment by beef9999** [comment]
arrogent github

**Comment by samskiter** [comment]
This still catches me out so often. The given explanation of threedot that it 'shows the changes introduced' is just wrong - the changes introduced are the diffs between the heads of the two branches.

I suspect this is a compute optimisation, not to have to recreate diffs when the target branch updates.

It might also make the 'changes since last review' view harder, but something like 'interdiff' is a tool to help solve this.

**Comment by EssKayz** [comment]
We today once again encountered this exact problem.
We were merging a PR with no conflicts, where a logger object was added in the PR branch.
However, the same logger object was already initialized in another feature branch in the same file.
This logger object was not visible in the PR diff, neither as a conflict nor as an existing object in the PR.
This directly resulted in the logger now being initialized twice once the PR was merged.
This has also happened to us previously in a much more sinister context, which directly caused a production outage by duplicating a few hundreds of lines of configurations of a PR merge.

![image](https://github.com/user-attachments/assets/d86a6e3f-7ea0-4fa1-8a9c-b230ad0222e1)
![image](https://github.com/user-attachments/assets/c15170f7-57d9-4b92-8708-66dc403f68ac)

**Reply by EssKayz** [reply]
This REALLY reduces trust in the PR "View files" page, since it does not do what the documentation says.
[Docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests)
"Rather than viewing the commits themselves, you can view the proposed changes as they'll appear in the files once the pull request is merged." 
^ That is simply untrue, the PR does not show the changes as they will appear, but rather how they appeared in the feature branch.

**Comment by cody-owens** [comment]
It's frankly bizarre that GitHub has no Pull Request two dot option, and has caused me to lose confidence in GitHub as a production-ready repository management vendor. We have run into various issues where, due to these three dot diffs, pull requests would show "phantom changes".

The other commenters here are correct - the GitHub docs description for three dot comparison is a lie. What a reviewer cares about is what real changes will enter the target branch if the PR is merged. If your Pull Request diffs view shows _anything_ other than _exactly that_, it is unreliable and strong a reason for a company to pick a different vendor capable of production-ready code review support.
